utils/s3/client.cc

```diff
@@ -854,6 +861,209 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
     return data_sink(std::make_unique<upload_jumbo_sink>(shared_from_this(), std::move(object_name), max_parts_per_piece));
 }
 
+// unlike upload_sink and upload_jumbo_sink, do_upload_file reads from the
+// specified file, and sends the data read from disk right away to the wire,
+// without accumulating them first.
+class client::do_upload_file {
+    shared_ptr<client> _client;
+    const std::filesystem::path _path;
+    sstring _object_name;
+    sstring _upload_id;
+    utils::chunked_vector<sstring> _part_etags;
```

Comment:
nit: Since we know the part size in advance we know the number of parts, thus we can .reserve() it prior to populating